[wip] haskell: add core libraries logic to remove need for global package db

matthewbauer opened the pull request:
This moves core libraries in Haskell to its own derivations. This is
needed to avoid references to the GHC compiler ending up in binaries
built by GHC. Instead of referencing GHC directly, these binaries will
now link to the copies.

Unfortunately, for this to work, we need to know which GHC libraries
exist at evaluation time. This requires maintaining a list of the
builtin library names & their versions in the GHC compiler’s “libs”
attribute. A better way of handling this is needed.

This is still a work in progress.

/cc @kirelagin @bgamari @ElvishJerricco

matthewbauer committed:
haskell: add core libraries logic

This moves core libraries in Haskell to its own derivations. This is
needed to avoid references to the GHC compiler ending up in binaries
built by GHC. Instead of referencing GHC directly, these binaries will
now link to the copies.

Unfortunately, for this to work, we need to know which GHC libraries
exist at evaluation time. This requires maintaining a list of the
builtin library names & their versions in the GHC compiler’s “libs”
attribute. A better way of handling this is needed.

peti commented:
How do these changes relate to https://github.com/NixOS/nixpkgs/pull/43070?

matthewbauer commented:
I think they accomplish the same thing - just in different ways. Both should reduce closure sizes.

Pros of this approach:
- Doesn't rely on multiple outputs.
- Avoids rebuilding GHC.
- Meant as a stopgap fix until we can get GHC to split up core libraries from the compiler.

Cons of this approach:
- You end up with two copies of each core library in your Nix store.
- Requires keeping track of name + version of every core library in the GHC compiler (more maintenance especially because it looks like they change from version to version)

matthewbauer committed:
haskell: use ghc861 for core libraries

ghc8.2.2 apparently doesn’t like moving rts from its original
location. Hopefully this was fixed in ghc 8.6.1.

matthewbauer commented:
This is currently blocked on an issue that I have been unable to debug:

```
Configuring hscolour-1.24.4...
Dependency base <10: using base-4.12.0.0
Dependency containers -any: using containers-0.6.0.1
Dependency base <10: using base-4.12.0.0
Dependency containers -any: using containers-0.6.0.1
<command line>: unknown package: rts
```

It looks like it picks up base but not rts?

matthewbauer committed:
haskell: put core libs in propagated build inputs

This may not be needed but it is more consistent with the idea of
treating core libraries as ordinary haskell libraries.

Ericson2314 commented:
@peti Unless we're worried about churn for the builders, I'd love to land #43070 as soon as possible, as it was passing deps before mere conflicts. Then we can do this approach, with the libraries having been listed for us, right after.

domenkozar commented:
This is really good work, it will reduce haskell closures significantly for packages dynamically linked.

kirelagin commented:
> Both should reduce closure sizes.

Actually, I don’t think that my approach affects closure size in any way.

kirelagin commented:
I might be missing something, but can’t we just “merge” our two PRs by moving the code from `mkCoreLib` here into the install phase in my PR? This way boot libraries outputs will not retain a link to the main output and we will get the closure size reduced.

P. S. I wonder if the GHC build system can be told to install those packages straight into the right location instead of us copying them around manually.

kirelagin commented:
> having a separate output for each library is kind of an abuse of the idea of multiple outputs

I don’t see how this is an abuse TBH. The GHC build process is structured in such a way that it produces both the compiler and libraries for it to use and, furthermore, it is the only supported way of getting those libraries. What I mean is that some of the boot libraries can be then recompiled, but some of them just cannot be built outside of the ghc compilation process, and, AFAIU, this is exactly the purpose of multiple outputs.

What we can do is limit those libraries that go to multiple outputs strictly to those that cannot be built separately.

Ericson2314 commented:
@kirelagin I personally don't mind multiple outputs, but I do think it's important that we have it so generic-builder and ghc with generic-builder can find them exactly one way. Whether we do that with multiple outputs, or shim derivations I guess doesn't matter as much on that front as long as GHC-without-libraries has its own output / shim derivation. Put another way, the `-B` for GHC should give it a thing without libraries.

matthewbauer commented:
> I don’t see how this is an abuse TBH. The GHC build process is structured in such a way that it produces both the compiler and libraries for it to use and, furthermore, it is the only supported way of getting those libraries. What I mean is that some of the boot libraries can be then recompiled, but some of them just cannot be built outside of the ghc compilation process, and, AFAIU, this is exactly the purpose of multiple outputs.

The original use case was for splitting up ```/include``` from ```/lib```. I don't think anyone has ever generated them from a variable list like this before. There's probably no technical reason why it can't be done, it is just a little bit unexpected. A lot of this can probably be solved in tooling. For instance there is no real way to "discover" which of these outputs exist (things like ```nix-env -qa``` won't work with multiple outputs).

kirelagin commented:
Well, those multiple outputs are just a technical detail, it is actually a _good_ thing that there is no easy way to discover them. They are intended to be used only as top-level haskell packages set attributes.

Ericson2314 commented:
Another concern is that more invasive the procedure gets, The move convenient it is to have two derivations for purpose of debug cycles. Kind of like how we don't wrap gcc when be build it but separately.

matthewbauer committed:
haskell: cleanup core libs

Move score libs to pkgs/development/haskell-modules/default.nix. This
seems like a better place for them.

Instead of doing “base = null;” we now can use the actual base here.

matthewbauer commented:
I honestly have no idea how to fix the ```unknown package: rts```. Any debugging points to it being some sort of Cabal thing. I filed an issue here: https://github.com/haskell/cabal/issues/5445. Does anyone else have experience with using ```-no-global-package-db```?

matthewbauer committed:
haskell: fix rts case

Ericson2314 commented on pkgs/development/compilers/ghc/8.6.1.nix:
```diff
@@ -208,6 +208,45 @@ stdenv.mkDerivation (rec {
 
     # Our Cabal compiler name
     haskellCompilerName = "ghc-8.4.3";
```
Unrelated issue here, btw

cdepillabout commented:
@matthewbauer This PR is quite old at this point, is there anything in particular that needs to be done to move it forward?

Or could we go ahead and close it?

matthewbauer commented:
Yes closing for now. It may be useful in the future but not in the current state
